jenkins/janitor.py

```diff
@@ -0,0 +1,151 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import collections
+import datetime
+import json
+import pytz
+import subprocess
+import sys
+
+from collections import namedtuple
+from dateutil import parser as dateparser
+
+
+# A resource that need to be cleared.
+Resource = namedtuple('Resource', 'name condition status')
+DEMOLISH_ORDER = [
+    # Beaware of insertion order
+    Resource("instances", "zone", None),
+    Resource("addresses", "region", None),
+    Resource("disks", "zone", None),
+    Resource("firewall-rules", None, None),
+    Resource("routes", None, None),
+    Resource("forwarding-rules", "region", None),
+    Resource("target-pools", "region", None),
+    Resource("instance-groups", "zone", "managed"),
+    Resource("instance-templates", None, None),
+]
+
+
+def collect(project, age, resource, filt):
+    """ Collect a list of resources for each condition (zone or region).
+
+    Args:
+        project: The name of a gcp project.
+        age: Time cutoff from the creation of a resource.
+        resource: Definition of a type of gcloud resource.
+        filt: Filter clause for gcloud list command.
+    Returns:
+        A dict of condition : list of resource.
+    """
+
+    col = collections.defaultdict(list)
+
+    for item in json.loads(subprocess.check_output([
+            'gcloud', 'compute', '-q',
+            resource.name, 'list',
+            '--format=json',
+            '--filter=%s' % filt,
+            '--project=%s' % project])):
+
+        if not item['name'] or not item['creationTimestamp']:
+            print "[Warning] - Skip item without name or timestamp"
+            print "%s" % item
+            continue
+        if resource.condition and not item[resource.condition]:
+            print "[Warning] - condition specified but not found"
+            continue
+
+        if resource.condition:
+            colname = item[resource.condition]
+        else:
+            colname = ""
+
+        # Unify datetime to use utc timezone.
+        created = dateparser.parse(item['creationTimestamp']).astimezone(pytz.utc)
+        print "Found %s, %s in %s" % (resource.name, item['name'], colname)
+        if created < age:
+            print "Include %s, %s" % (resource.name, item['name'])
+            col[colname].append(item['name'])
+    return col
+
+
+def clear_resources(project, col, resource):
+    """Clear a collection of resource, from collect func above.
+
+    Args:
+        project: The name of a gcp project.
+        col: A dict of collection of resource.
+        resource: Definition of a type of gcloud resource.
+    Returns:
+        0 if no error
+        1 if deletion command fails
+    """
+    err = 0
```

Comment:
nit: this `err` idiom isn't pythonic. I think it's fine to leave for now, though.